core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1188,119 +1133,201 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
+      case(leaderBroker, topicAndPartitionsForBroker) => {
+        var imbalanceRatio: Double = 0
+        var topicsNotInPreferredReplica: Map[TopicAndPartition, Seq[Int]] = topicAndPartitionsForBroker
+          .filter { case (topicPartition, _) =>
+          controllerContext.partitionLeadershipInfo.contains(topicPartition) &&
+            controllerContext.partitionLeadershipInfo(topicPartition).leaderAndIsr.leader != leaderBroker
+        }
+        debug("topics not in preferred replica " + topicsNotInPreferredReplica)
+        val totalTopicPartitionsForBroker = topicAndPartitionsForBroker.size
+        val totalTopicPartitionsNotLedByBroker = topicsNotInPreferredReplica.size
+        imbalanceRatio = totalTopicPartitionsNotLedByBroker.toDouble / totalTopicPartitionsForBroker
+        trace("leader imbalance ratio for broker %d is %f".format(leaderBroker, imbalanceRatio))
+        // check ratio and if greater than desired ratio, trigger a rebalance for the topic partitions
+        // that need to be on this broker
+        if (imbalanceRatio > (config.leaderImbalancePerBrokerPercentage.toDouble / 100)) {
+          topicsNotInPreferredReplica.keys.foreach { topicPartition =>
+            // do this check only if the broker is live and there are no partitions being reassigned currently
+            // and preferred replica election is not in progress
+            if (controllerContext.liveBrokerIds.contains(leaderBroker) &&
+                controllerContext.partitionsBeingReassigned.isEmpty &&
+                controllerContext.partitionsUndergoingPreferredReplicaElection.isEmpty &&
+                !topicDeletionManager.isTopicQueuedUpForDeletion(topicPartition.topic) &&
+                controllerContext.allTopics.contains(topicPartition.topic)) {
+              onPreferredReplicaElection(Set(topicPartition), true)
             }
           }
         }
       }
     }
   }
-}
 
-/**
- * Starts the partition reassignment process unless -
- * 1. Partition previously existed
- * 2. New replicas are the same as existing replicas
- * 3. Any replica in the new set of replicas are dead
- * If any of the above conditions are satisfied, it logs an error and removes the partition from list of reassigned
- * partitions.
- */
-class PartitionsReassignedListener(protected val controller: KafkaController) extends ControllerZkDataListener {
-  private val controllerContext = controller.controllerContext
+  def getControllerID(): Int = {
+    controllerContext.zkUtils.readDataMaybeNull(ZkUtils.ControllerPath)._1 match {
+      case Some(controller) => KafkaController.parseControllerId(controller)
+      case None => -1
+    }
+  }
 
-  protected def logName = "PartitionsReassignedListener"
+  class ControllerEventThread(name: String) extends ShutdownableThread(name = name) {
+    override def doWork(): Unit = {
+      val controllerEvent = controllerEventQueue.take()
+      try {
+        controllerEvent.process()
```

Comment:
I'm also a bit uncomfortable with manually doing the check in every event. It seems easy to introduce problems when adding new events or changing existing ones. If reelecting the controller is really the only case that needs special treatment, then Jun's suggestion makes sense.

Otherwise, we could have a `processIfActive` method in `ControllerEvent` and by default `process` would do the `isActive` check, but `process` could be overridden if necessary.